reasoner/resolution/resolver/ConcludableResolver.java

```diff
@@ -92,101 +81,73 @@ public void receiveRequest(Request fromUpstream, int iteration) {
-        if (requestState.isExploration() && requestState.asExploration().singleAnswerRequired()) {
-            requestState.asExploration().downstreamManager().clearDownstreams();
-            requestState.answerCache().setComplete();
+            answerToUpstream(fromDownstream.answer(),
+                             requestMapByRoot.get(fromDownstream.answer().root()).get(fromDownstream.sourceRequest()),
+                             iteration);
         }
-        answerToUpstream(upstreamAnswer, fromUpstream, iteration);
     }
 
     @Override
-    protected void receiveFail(Response.Fail fromDownstream, int iteration) {
-        LOG.trace("{}: received Fail: {}", name(), fromDownstream);
-        if (isTerminated()) return;
-
-        Request toDownstream = fromDownstream.sourceRequest();
-        Request fromUpstream = fromUpstream(toDownstream);
-
-        if (iteration < iterationByRoot.get(fromUpstream.partialAnswer().root())) {
+    protected void receiveFail(Response.Fail fail, int iteration) {
+        if (iteration < iterationByRoot.get(fail.sourceRequest().partialAnswer().root())) {
             // short circuit old iteration failed messages to upstream
-            failToUpstream(fromUpstream, iteration);
+            failToUpstream(fail.sourceRequest(), iteration);
         } else {
-            CachingRequestState<?, ConceptMap> requestState = this.requestStates.get(fromUpstream);
-            assert iteration == requestState.iteration();
-            if (requestState.isExploration()) {
-                requestState.asExploration().downstreamManager().removeDownstream(fromDownstream.sourceRequest());
-            }
-            nextAnswer(fromUpstream, requestState, iteration);
+            Request request = fail.sourceRequest();
+            subsumptionTrackers
+                    .computeIfAbsent(request.partialAnswer().root(), r -> new SubsumptionTracker())
+                    .addFinished(request.partialAnswer().conceptMap());
+            failToUpstream(requestMapByRoot.get(fail.sourceRequest().partialAnswer().root()).get(request),
+                           iteration);
         }
     }
 
-    public void receiveReiterationQuery(ReiterationQuery.Request request) {
-        boolean reiterate = cacheRegistersByRoot.containsKey(request.sender()) &&
-                iterate(cacheRegistersByRoot.get(request.sender()).values())
-                    .filter(AnswerCache::requiresReiteration).first().isPresent();
-        driver().execute(actor -> actor.respondToReiterationQuery(request, reiterate));
-    }
-
-    private void respondToReiterationQuery(ReiterationQuery.Request request, boolean reiterate) {
-        request.onResponse().accept(ReiterationQuery.Response.create(driver(), reiterate));
-    }
-
-    @Override
-    public void terminate(Throwable cause) {
-        super.terminate(cause);
-        requestStates.clear();
-    }
-
     @Override
     protected void initialiseDownstreamResolvers() {
```

Comment:
note: if we don't need `resolverRules` in `BoundConcludable` make sure you get rid of it here too, only used for that purpose